### PR TITLE
[needs-docs] update metadata viewer and review UI for vector layer dialog

### PR DESCRIPTION
**Pull request by Gustry**
## Description

This is part of the [QEP 91 work package 4](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/91) about metadata viewer.

Previously: 
* In the general tab, some fields were read-only : source and display name and they were related to metadata.
screen shot 2017-03-30 at 14 46 53

* In the metadata tab, the HTML read-only output was mixed with editable fields
screen shot 2017-03-30 at 14 47 29

I suggest to have: 
* an information tab, read only to display information and metadata about the layer. I tried to have something similar with Geonode: http://188.165.62.190/layers/geonode:italy/metadata_detail
* the metadata tab, the same as before, but with layer name as well and without the HTML output.
* a source tab, to be able to edit provider filter, scale, spatial index

Preview
* Information

screen shot 2017-04-12 at 16 04 29

* Metadata

screen shot 2017-03-29 at 13 56 59

* Source
screen shot 2017-04-12 at 14 37 19

* Rendering

screen shot 2017-04-12 at 14 37 29

The information tab will display more information after the work from @nyalldawson about  [QEP 91 work package 2](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/91) : 
* contacts
* categories, language, date ...

This is the first pass about this QEP implementing the metadata viewer.
I will come back on the metadata editor later when we will have work package 2 implemented.

BTW : For string concatenation (for HTML), in QT5 we can use QStringBuilder. I didn't find if it was enabled by default for QGIS. I found some references in the code. See http://doc.qt.io/qt-5/qstring.html#more-efficient-string-construction

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [x] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [ ] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [x] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [ ] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [ ] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

Funder : 
 * WorldBank
 * Kartoza

CC : @timlinux

### CODE REVIEW

**Comment by timlinux** [comment]
👍  Lovely stuff @Gustry

**Comment by glaroc** [comment]
Great idea to move the CRS menu to a Source section. I would even add a label that makes it clear that this doesn't "update" the CRS of the layer, but "specifies it" in the event that it's not or improperly specified. When I teach QGIS workshops, this is the single most common and worst mistake made by participants. They change the CRS in the Layer properties dialog thinking that it will update/change it, leaving them with mis-specified CRS. 

I assume the Spatial field in the Information section will also show the resolution and number of rows and columns for rasters? If the CRS could be shown with the short name as well EPSG:4326 (Geographic: WGS 84), that would be great.

**Comment by Gustry** [comment]
> Great idea to move the CRS menu to a Source section.

Agree, it will be better to make the user understand about source CRS. I will update the label too, good idea.

> I assume the Spatial field in the Information section will also show the resolution and number of rows and columns for rasters?

Yes, this is just the mockup for vector layer. I will do the similar approach for raster when this will be approved.

>  If the CRS could be shown with the short name as well EPSG:4326 (Geographic: WGS 84), that would be great.

Done, thanks

**Comment by dakcarto** [comment]
@Gustry nice separation of content. Btw, the `displayed as` name is an artifact from how a layer's name has been changed by other QGIS app settings, i.e. how it appears in the Layers Panel, specifically whether `Capitalize layer names` is checked under the `Canvas & Legend` section of app options.

So, it is not metadata-relative, per se. Not sure where it should go, but seems OK in Metadata to me.

Also, I like the idea of defining `QT_USE_QSTRINGBUILDER`; seems like an interestingly simple way to speed things up.

File: src/core/qgsvectorlayer.h [hunk 1]
```diff
@@ -39,6 +40,7 @@
 #include "qgsaggregatecalculator.h"
 #include "qgsfeatureiterator.h"
 #include "qgsexpressioncontextgenerator.h"
+#include "qgsunittypes.h"
```
Comment (m-kuhn) [inline src/core/qgsvectorlayer.h]: If not strictly required to `#include ` headers in the header, do it in the source file. I.e. move these includes here to qgsvectorlayer.cpp

**Comment by giohappy** [comment]
I agree with @glaroc on adding a informative text for the CRS setting. This is the most common I face in my workshops too, and this means it's not a sufficiently clear UI.

**Comment by nyalldawson** [comment]
+1 from me, with one request:

Can the "scale based rendering" widget be moved from the new "source" tab to the "rendering" tab? It really doesn't fit well under "source".

I'm not sure how to handle this for raster layers though - there's not really a good counterpart to vector layers "rendering" tab....

**Comment by Gustry** [comment]
> I agree with @glaroc on adding a informative text for the CRS setting. 

Do you have a better wording for this? Having the CRS combobox in the `source` tab make it a little bit clearer.

> Can the "scale based rendering" widget be moved from the new "source" tab to the "rendering" tab?

Sure, will do that now. 

> I'm not sure how to handle this for raster layers though

I wanted to do the same approach. But I agree, the metadata HTML output is quite different now in a QgsRasterLayer.

**Comment by nyalldawson** [comment]
> I wanted to do the same approach. But I agree, the metadata HTML output is quite different now in a QgsRasterLayer.

Ah ok - so currently no changes to raster layers?

**Comment by Gustry** [comment]
In this PR, no nothing. Is-it a blocker to have this PR merged?
I wanted first to a have some feedbacks about this new user experience for the vector layer properties.

**Comment by nyalldawson** [comment]
Not a blocker in my opinion

**Comment by Gustry** [comment]
I have updated my PR and my screenshots in the description.
Does it look good for you ?

**Comment by m-kuhn** [comment]
Cool.
Can you add some more spacing in the Information (HTML?) part between the title and the content (Between `ID` and `small_grid_...`)

**Comment by Gustry** [comment]
There is not a lot of supported CSS, but it's done @m-kuhn 
I updated the screenshot.

**Comment by timlinux** [comment]
I am going to go ahead and merge this folks - hope you are all OK with that - there is more coming soon.

**Pull request merged**